lib/src/generator_frontend.dart

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io' show File;
+
+import 'package:dartdoc/src/generator.dart';
+import 'package:dartdoc/src/logging.dart';
+import 'package:dartdoc/src/model/model.dart';
+import 'package:dartdoc/src/model_utils.dart';
+import 'package:dartdoc/src/warnings.dart';
+import 'package:path/path.dart' as path;
+
+typedef FileWriter = File Function(String filePath, Object content,
+    {bool allowOverwrite, Warnable element});
+
+/// [Generator] that delegates rendering to a [GeneratorBackend] and delegates
+/// file creation to a [FileWriter].
+class GeneratorFrontEnd implements Generator {
+  final GeneratorBackend _generatorBackend;
+  final FileWriter _writer;
+
+  // Used in write(). This being not null signals the generator is active.
+  String _outputDirectory;
+
+  final StreamController<void> _onFileCreated = StreamController(sync: true);
+
+  @override
+  Stream<void> get onFileCreated => _onFileCreated.stream;
+
+  @override
+  final Map<String, Warnable> writtenFiles = {};
+
+  GeneratorFrontEnd(this._generatorBackend, this._writer);
+
+  // Implement FileWriter so we can wrap the one given to us.
+  File write(String filePath, Object content,
+      {bool allowOverwrite, Warnable element}) {
+    assert(_outputDirectory != null);
+    // Replace '/' separators with proper separators for the platform.
+    String outFile = path.joinAll(filePath.split('/'));
+    outFile = path.join(_outputDirectory, outFile);
+    File file = _writer(outFile, content,
+        allowOverwrite: allowOverwrite, element: element);
+    writtenFiles[outFile] = element;
```

Comment:
Build fails because of `outFile` here, which causes the orphan check to fail to ignore `categories.json` and `__404error.html`. Using `filePath` (which doesn't include the `outputDirectory` part) restores the original behavior, but not sure if that is the right thing to do here.

Reply:
The orphan check detects ignores via substrings of the relative path.  That could be changed, and I'd support that.

Reply:
Should be fixed in latest update